README.md
# ApiClient

[![Gem Version](https://badge.fury.io/rb/api-client.png)](http://badge.fury.io/rb/api-client) [![Build Status](https://secure.travis-ci.org/zertico/api-client.png?branch=master)](http://travis-ci.org/zertico/api-client) [![Dependency Status](https://gemnasium.com/zertico/api-client.png)](https://gemnasium.com/zertico/api-client) [![Coverage Status](https://coveralls.io/repos/zertico/api-client/badge.png?branch=master)](https://coveralls.io/r/zertico/api-client)  [![Code Climate](https://codeclimate.com/github/zertico/api-client.png)](https://codeclimate.com/github/zertico/api-client)

ApiClient handle all the logic necessary to call Some API, catch the response and initialize an object with it for you.
It is possible to use Typhoeus or the native Ruby Library Net::Http.
It works very well with rails, so you can exchange your ActiveRecord based models without any concern and your application will make Api calls transparently.
It supports ruby 1.8.7, 1.9.2, 1.9.3, jruby and ree out of the box.

## Useful Links

### Mailing List

If you have any questions, comments, or concerns, please use the Google Group instead of the Github issue tracker:

[https://groups.google.com/forum/#!forum/zertico-api-client](https://groups.google.com/forum/#!forum/zertico-api-client)

### Documentation

You can find the code documentation for the last version generated by Yard on this link:

[http://rdoc.info/gems/api-client/frames](http://rdoc.info/gems/api-client/frames)

### Changelog

[https://github.com/zertico/api-client/blob/master/CHANGELOG.md](https://github.com/zertico/api-client/blob/master/CHANGELOG.md)

## Installation

Add this line to your application's Gemfile:

    gem 'api-client'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install api-client

If you will use Typhoeus (https://github.com/typhoeus/typhoeus), you must have Typhoeus version above 0.5.0.

## Basic Usage

Create an initializer:

```ruby
ApiClient.configure do |config|
   # You can define an api entry point
   config.path = 'http://api.example.com'
   # or several ones
   config.paths = { :default => 'http://api.example.com', :auth => 'http://auth.example.com' }
   # Default header
   config.header = { 'param1' => '123329845729384759237592348712876817234'}
   # Basic Auth
   config.basic_auth('user', 'pass')
   # If inside Rails
   config.mock = Rails.env.test?
end
```

Add this to your ApplicationController:

```ruby
rescue_from ApiClient::Exceptions::NotFound, :with => :not_found

def not_found
    #Specify your own behavior here
end
```

You can define a more generic rescue that will work for any error:

```ruby
rescue_from ApiClient::Exceptions::Generic, :with => :generic_error
```

On Your model, extend ApiClient::Base

```ruby
class User < ApiClient::Base
```

Then, on your action, just put into it:

```ruby
@user = User.get(3)
#or
@user = User.find(3)
```

where 3 is the id of the user.

To a request that returns a collection of the object, use:

```ruby
@user = User.collection
#or
@user = User.all
```

## Advanced Usage

ApiClient can read api responses with root nodes based on the name of the virtual class.
In Some cases, that is not the required behavior. To Redefine it, use remote_object method:

```ruby
class Admin < ApiClient::Base
    self.root_node = 'user'
end
```

To specify a resource path different from the expected, you can overwrite the behavior by setting resouce_path:

```ruby
class Admin < ApiClient::Base
    self.resource_path = 'users?type=admin'
end
```

It can handle associations. It will automatically instantiate an association for you if properly setted like below:

```ruby
class Person < ApiClient::Base
    self.associations = { :houses => "House", :cars => "Car" }
end
```

In case you need to work with one api entry point, it will define you path as the default.
If you need multiple entry points, you must define a name to each one, so you can refer to them on the model as:

```ruby
class User < ApiClient::Base
    self.path = :auth
end
```

This code will create a setter and a getter for houses and cars and initialize the respective class.

When you are working with collections you can use api pagination according Hal Specification (http://stateless.co/hal_specification.html) .
You just need to pass a hash as:

```ruby
{
  total: 10,
  total_pages: 1,
  offset: 10,
  _links: {
    first: { href: "/api/clients" },
    previous: null,
    self: { href: "/api/clients" },
    next: null,
    last: { href: "/api/clients?page=1" }
  },
  users: [ { user: { name: "example1" } }, { user: { name: "example2" } } ]
}
```


Since version 2.0.0, it is possible to make api calls from the object. The syntax is pretty much the same.
It will make the call and update the fields with the response. Look at the examples folder to see more code examples

## Testing

Set mock config variable as true inside yout spec_helper or test_helper:

```ruby
ApiClient.configure do |config|
   config.mock = true
end
```

With this setting no requisitions will be made. All calls will just return a new object with the attributes received.

## Parallel

When making parallel requests, the requests are made in threads, so to get the response it is necessary to specify
an initialized object to update when the requisition is complete.

```ruby
@users = ApiClient::Collection.new({}, User)
@cars = ApiClient::Collection.new({}, Car)
@house = House.new

ApiClient.parallel do
    User.all.on_complete_update(@users)
    Car.all.on_complete_update(@cars)
    House.find(1).on_complete_update(@house)
end
```

## Migrating from 2.\*.\* to 3.0.0

Since version 3.0.0 is not necessarily set the root_node on the attributes when calling methods on the class.

Before:

```ruby
@user = User.create({ :user => attributes })
```

After:

```ruby
@user = User.create(attributes )
```

## More Examples
[Project](https://github.com/zertico/api-client/tree/master/examples)

## TODO
* Add more Response Handlers

## Mantainers
[@plribeiro3000](https://github.com/plribeiro3000)

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
# CHANGELOG

## v3.1.0

* Setting Default Accept header as json

## v3.0.0

* Automatically sending attributes with root_node when calling methods on class.

## v2.7.0

* Fixed update routes to add resource id.

## v2.6.0

* Add support for pagination attributes.

## v2.5.0

* Add support for parallel requests. Yay! =)

## v2.4.0

* Add support to multiple api entry points.
* Added license to gemspec.

## v2.3.0

* Add support to global mock of requisitions for testing purposes.

## v2.2.0

* Add support global basic authentication.
* Add support for gem Zertico

## v2.1.0

* Add support for classes with compost names.

## v2.0.3

* Fix put and patch requisitions.

## v2.0.2

* Changed to_hash method behavior to return a root node.
* Fix instance requisitions by send a root node.

## v2.0.1

* Removed trailing slash from post requisitions.

## v2.0.0

* Fix typhoeus requisitions by changing params method to body.
* Add Support to configure header globally.
* Improved readability by changing methods names.

## v2.0.0.rc2

* Fix initialization with root nodes for nested objects.
* Add Support for object initialization with strings as keys.

## v2.0.0.rc1

* Add Support for global url.
* Add Support for instance requests.
* Changed behavior of request methods to only accepts the resource id if needed.

## v1.10.0

* Add Support for api responses with collections.

## v1.9.1

* Fix for Net::Http on NotFound Exception.

## v1.9.0

* Added Requested Url to NotFound Exception message.

## v1.8.2

* Fixed Query String Bug (It was been sent empty).

## v1.8.1

* Fixed a Bug when response was an Array.

## v1.8.0

* Methods to return attributes of the class added.

## v1.7.0

* Support for Associations added.

## v1.6.0

* Initial Support for Typhoeus added.
* Fixed bug with get request and headers.
* Support for ruby 1.8.7, 1.9.2, jruby and ree added.

## v1.5.1 to v1.5.4

* Fixed Bug with errors from external API and self.

## v1.5.0

* Handler of arrays as response.
* Method to redefine response object added.

## v1.4.1

* ActiveModel compatibility for errors added. make all error keys symbolic.

## v1.4.0

* functionality to validate on client side using ApiClient::Base added.

## v1.3.0

* functionality to read args with a root node added.

## v1.2.0

* methods put, patch and delete added.
* changed post method to send as json by default, but left option to override it.
* header params added to methods.

## v1.1.1

* fix for post when uri has a non default port (80).

## v1.1.0

* changed post method to always send as json.

## v1.0.0

* handler for response code 400 removed.
* Refactoring of usage.

## v0.6.0

* handler for response code 400 added.

## v0.5.0

* handler for connection refused added.

## v0.4.1

* rename gem name to fix require issues.

## v0.4.0

* handlers for response code 500, 502 and 503 added.

## v0.3.0

* handler for response code 401 added.

## v0.2.0

* post method added

## v0.1.0

* generic exception added.

## v0.0.1

* get method added.
* handler for response code 404 added.


lib/api-client/parser.rb
require 'json'

# ApiClient::Parser provides a method to parse the request response.
module ApiClient::Parser
  # Parse the JSON response.
  #
  # @param [HTTP] response HTTP object for the request.
  # @param [String] url The url of the requisition.
  # @return [Hash] the body parsed.
  def self.response(response, url)
    raise_exception(response, url)
    begin
      object = ::JSON.parse(response.body)
    rescue ::JSON::ParserError, TypeError
      object = {}
    end
    object
  end

  protected

  def self.raise_exception(response, url)
    case response.code.to_i
      when 401 then raise ApiClient::Exceptions::Unauthorized
      when 403 then raise ApiClient::Exceptions::Forbidden
      when 404 then raise ApiClient::Exceptions::NotFound.new(url)
      when 500 then raise ApiClient::Exceptions::InternalServerError
      when 502 then raise ApiClient::Exceptions::BadGateway
      when 503 then raise ApiClient::Exceptions::ServiceUnavailable
      else return
    end
  end
end